File: R/hks.R
library(tidyverse)
library(progress)
library(stats)
library(kSamples)

# --------------------------------- HKS/AD TEST ---------------------------------

hierarchical_permutation_test <- function(df,
                                          value_col,
                                          subject_col,
                                          cell_col,
                                          group_col,
                                          metric = "ks",
                                          n_cells_per_subject = NULL,
                                          n_obs_per_cell = NULL,
                                          n_resamples = 1000,
                                          n_permutations = 1000,
                                          replace = FALSE,
                                          seed = NULL) {
  
  
  if (!is.null(seed)) set.seed(seed)
  df <- as_tibble(df)
  
  # --------------------------------- 1. Group Validation ---------------------------------
  
  groups <- unique(df[[group_col]])
  if (length(groups) != 2) {
    stop(sprintf("The column '%s' must contain exactly 2 groups.", group_col))
  }
  
  # --------------------------------- 2. Setting Number of Observations ---------------------------------
  
  actual_min_obs <- df %>%
    group_by(!!sym(subject_col), !!sym(cell_col)) %>%
    count() %>%
    pull(n) %>%
    min()
  
  if (is.null(n_obs_per_cell)) {
    n_obs <- actual_min_obs
    message(sprintf("INFO: Auto-detected n_obs_per_cell = %d", n_obs))
  } else {
    if (!replace && n_obs_per_cell > actual_min_obs) {
      stop(sprintf("ERROR: Requested %d observations, but min available is %d. Set 'replace = TRUE'.", 
                   n_obs_per_cell, actual_min_obs))
    } else if (replace && n_obs_per_cell > actual_min_obs) {
      warning(sprintf("WARNING: Requested %d observations (available: %d). Oversampling used.", 
                      n_obs_per_cell, actual_min_obs))
    }
    
    n_obs <- n_obs_per_cell
  }
  
  # --------------------------------- 3. Setting Number of Cells ---------------------------------
  
  actual_min_cells <- df %>%
    group_by(!!sym(subject_col)) %>%
    summarise(n_unique = n_distinct(!!sym(cell_col)), .groups = "drop") %>%
    pull(n_unique) %>%
    min()
  
  if (is.null(n_cells_per_subject)) {
    n_cells <- actual_min_cells
    message(sprintf("INFO: Auto-detected n_cells_per_subject = %d", n_cells))
  } else {
    if (!replace && n_cells_per_subject > actual_min_cells) {
      stop(sprintf("ERROR: Requested %d cells, but min available is %d. Set 'replace = TRUE'.", 
                   n_cells_per_subject, actual_min_cells))
    } else if (replace && n_cells_per_subject > actual_min_cells) {
      warning(sprintf("WARNING: Requested %d cells (available: %d). Oversampling used.", 
                      n_cells_per_subject, actual_min_cells))
    }
    
    n_cells <- n_cells_per_subject
  }
  
  # --------------------------------- 4. Resampling and Test ---------------------------------
  
  hierarchical_resample <- function(data) {
    
    # Observations
    balanced_obs <- data %>%
      group_by(!!sym(subject_col), !!sym(cell_col)) %>%
      slice_sample(n = n_obs, replace = replace) %>%
      ungroup()
    
    # Cells
    selected_cells <- balanced_obs %>%
      select(!!sym(subject_col), !!sym(cell_col)) %>%
      distinct() %>%
      group_by(!!sym(subject_col)) %>%
      slice_sample(n = n_cells, replace = replace) %>%
      ungroup
  
    # Merge
    inner_join(balanced_obs, 
               selected_cells, 
               by = c(subject_col, cell_col),
               relationship = "many-to-many")
  }
  
  compute_stat <- function(data) {
    g1 <- data[[value_col]][data[[group_col]] == groups[1]]
    g2 <- data[[value_col]][data[[group_col]] == groups[2]]
    if (metric == "ks") {
      return(suppressWarnings(ks.test(g1, g2)$statistic))
    } else if (metric == "ad") {
      return(suppressWarnings(kSamples::ad.test(g1, g2)$ad[1, 1]))
    }
  }
  
  # --------------------------------- 5. Bootstrap --------------------------------- 
  
  message("Calculating observed distribution...")
  pb1 <- progress_bar$new(total = n_resamples)
  obs_stats <- replicate(n_resamples, {
    pb1$tick()
    resampled <- hierarchical_resample(df)
    compute_stat(resampled)
  })
  
  obs_median <- median(obs_stats)
  
  message("Calculating null distribution...")
  subjects <- unique(df[[subject_col]])
  subject_labels <- df %>%
    select(all_of(c(subject_col, group_col))) %>%
    distinct()
  
  pb2 <- progress_bar$new(total = n_permutations)
  null_stats <- replicate(n_permutations, {
    pb2$tick()
    
    shuffled_labels <- sample(subject_labels[[group_col]])
    mapping <- setNames(shuffled_labels, subject_labels[[subject_col]])
    
    permuted_df <- df 
    permuted_df[[group_col]] <- mapping[as.character(permuted_df[[subject_col]])]
    
    resampled_perm <- hierarchical_resample(permuted_df)
    compute_stat(resampled_perm)
  })
  
  p_val <- mean(null_stats >= obs_median)
  
  return(list(
    p_value = p_val,
    obs_median = obs_median,
    obs_dist = obs_stats,
    null_dist = null_stats,
    metric = metric
  ))
}

# --------------------------------- PLOTTING ---------------------------------

plot_hks_results <- function(result_list, title = "Hierarchical Permutation Test") {
  
  df_plot <- data.frame(
    stat = c(result_list$null_dist, result_list$obs_dist),
    type = factor(c(rep("Null Distribution (Permuted)", length(result_list$null_dist)), 
                    rep("Observed Distribution (Resampled)", length(result_list$obs_dist))),
                  levels = c("Null Distribution (Permuted)", "Observed Distribution (Resampled)"))
  )
  
  metric_label <- if(is.null(result_list$metric)) "KS Statistic" else paste(toupper(result_list$metric), "Statistic")
  
  ggplot(df_plot, aes(x = stat, fill = type)) +

    geom_histogram(aes(y = after_stat(density)), 
                   alpha = 0.3, bins = 60, position = "identity", color = "black") +

    geom_vline(aes(xintercept = result_list$obs_median, color = "Observed Median"), 
               linetype = "dashed", linewidth = 1, show.legend = TRUE) +
    
    scale_fill_manual(values = c("Null Distribution (Permuted)" = "black", 
                                 "Observed Distribution (Resampled)" = "royalblue")) +
    scale_color_manual(name = "", values = c("Observed Median" = "red")) +
    
    labs(title = title, 
         subtitle = sprintf("p-value: %.5f", result_list$p_value),
         x = metric_label,
         y = "Density",
         fill = "Distributions") +
    
    theme_minimal() +
    theme(
      plot.title = element_text(hjust = 0.5, face = "bold", size = 14),
      plot.subtitle = element_text(hjust = 0.5, size = 12),           
      panel.border = element_rect(colour = "black", fill = NA, linewidth = 1), 
      axis.line = element_line(colour = "black"), 
      legend.position = "bottom"
    )
}

# --------------------------------- SIMULATED DATA ---------------------------------

generate_hierarchical_data <- function(n_basal,
                                       n_treatment,
                                       params_basal,
                                       params_treatment) {
  
  
  conditions <- list(
    list(name = "B", n = n_basal, params = params_basal),
    list(name = "T", n = n_treatment, params = params_treatment)
  )
  
  simulated_df <- map_df(conditions, function(cond) {
    map_df(seq_len(cond$n), function(s_idx) {
      subject_id <- paste0("S", cond$name, s_idx - 1)
      n_cells <- sample(10:60, 1)
      
      map_df(seq_len(n_cells), function(c_idx) {
        cell_id <- paste0("CELL_", subject_id, "_", c_idx - 1)
        n_samples <- sample(25:100, 1)
        samples <- rlnorm(n = n_samples,
                          meanlog = cond$params$mean,
                          sdlog = cond$params$sigma)
        
        tibble(
          subject = subject_id,
          subunit = cell_id,
          group = cond$name,
          variable = samples
        )
      })
    }) 
  })
  
  return(simulated_df)
}

# --------------------------------- TESTING ---------------------------------

# Generating the simulated dataframe
p_basal <- list(mean = 0.5, sigma = 0.3)
p_treatment <- list(mean = 0.5, sigma = 0.35)

df1 <- generate_hierarchical_data(
  n_basal = 12,
  n_treatment = 8,
  params_basal = p_basal,
  params_treatment = p_treatment
)

# Performing the HKS test
hks_result <- hierarchical_permutation_test(df = df1,
                                            value_col = 'variable',
                                            subject_col = 'subject',
                                            cell_col = 'subunit',
                                            group_col = 'group',
                                            metric = 'ad',
                                            n_resamples = 2000,
                                            n_permutations = 2000,
                                            replace = TRUE,
                                            seed = 123)

plot_hks_results(result_list = hks_result)

# Same test but with the AD metric
had_result <- hierarchical_permutation_test(df = df1,
                                            value_col = 'variable',
                                            subject_col = 'subject',
                                            cell_col = 'subunit',
                                            group_col = 'group',
                                            metric = 'ad',
                                            n_resamples = 2000,
                                            n_permutations = 2000,
                                            replace = TRUE,
                                            seed = 123)

plot_hks_results(result_list = had_result)

File: src/hks.py
import numpy as np
import pandas as pd
from scipy import stats
import matplotlib.pyplot as plt
import seaborn as sns
from tqdm import tqdm
from joblib import Parallel, delayed
import warnings
from typing import Literal, Optional, List, Dict

class HierarchicalPermutationTest:
    
    """
    Performs a hierarchical permutation test to compare the distributions of a nested 
    variable between two groups, using the Kolmogorov-Smirnov or Anderson-Darling statistic.
    This procedure handles nested data structures where subjects from two groups 
    (e.g., WT and Treatment) contain multiple measurement units (e.g., cells), 
    each yielding multiple distinct observations.
    
    Strategies for handling data structure:
    
    1. Hierarchical Resampling: Balances the weight of every cell and subject to
    mitigate the effects of unbalanced sample sizes (pseudoreplication).
    2. Subject-Level Permutation: Shuffles group labels at the subject level rather 
    than the observation level, strictly preserving intra-subject correlation.
    
    """
    
    
    def __init__(self,
                 df: pd.DataFrame,
                 value_col: str,
                 subject_col: str,
                 cell_col: str,
                 group_col: str,
                 metric: Literal['ks', 'ad'] = 'ks',
                 n_cells_per_subject: Optional[int] = None,
                 n_obs_per_cell: Optional[int] = None,
                 replace: bool = False):
        
        self.metric = metric
        self.replace = replace 
        
        self.groups = df[group_col].unique()
        if len(self.groups) != 2:
            raise ValueError(f"The column '{group_col}' must contain exactly 2 groups.")
        
        # ------------------------ 1. Setting the number of observations per cell ------------------------
        
        actual_min_obs = df.groupby([subject_col, cell_col])[value_col].count().min()
      
        if n_obs_per_cell is None:
            self.n_obs = actual_min_obs
            print(f"INFO: Auto-detected n_obs_per_cell = {self.n_obs} (minimum number across the entire dataset).")
        else:
            if not self.replace and n_obs_per_cell > actual_min_obs:
                raise ValueError(f"ERROR: Requested {n_obs_per_cell} observations, but the minimum available is {actual_min_obs}. "
                                 f"Reduce the number or set 'replace = True'.")
            elif self.replace and n_obs_per_cell > actual_min_obs:
                print(f"WARNING: Requested {n_obs_per_cell} observations (available: {actual_min_obs}). Oversampling (with replacement) will be used.")
            
            self.n_obs = n_obs_per_cell
            print(f"INFO: User defined n_obs_per_cell = {self.n_obs}.")
        
        # ------------------------ 2. Setting the number of cells per subject ------------------------
        
        actual_min_cells = df.groupby(subject_col)[cell_col].nunique().min()
      
        if n_cells_per_subject is None:
            self.n_cells = actual_min_cells
            print(f"INFO: Auto-detected n_cells_per_subject = {self.n_cells} (minimum number across the entire dataset).")
        else:
            if not self.replace and n_cells_per_subject > actual_min_cells:
                raise ValueError(f"ERROR: Requested {n_cells_per_subject} cells, but the minimum available is {actual_min_cells}. "
                                 f"Reduce the number or set 'replace = True'.")
            elif self.replace and n_cells_per_subject > actual_min_cells:
                print(f"WARNING: Requested {n_cells_per_subject} cells (available: {actual_min_cells}). Oversampling (with replacement) will be used.")

            self.n_cells = n_cells_per_subject
            print(f"INFO: User defined n_cells_per_subject = {self.n_cells}.")
            
        # ------------------------ 3. Data preprocessing ------------------------
        
        print("INFO: Optimizing data structure for fast resampling...")
        self.optimized_data = []
        self.subject_ids = df[subject_col].unique()
        self.group_map = {self.groups[0]: 0, self.groups[1]: 1}
        
        for sub in self.subject_ids:
            sub_df = df[df[subject_col] == sub]
            group_label = self.group_map[sub_df[group_col].iloc[0]]
            
            cells_data = []
            for cell in sub_df[cell_col].unique():
                vals = sub_df[sub_df[cell_col] == cell][value_col].to_numpy()
                cells_data.append(vals)
                
            self.optimized_data.append((group_label, cells_data))
            
        self.n_subjects = len(self.optimized_data)
        self.observed_distribution = []
        self.null_distribution = []
        self.p_value = None 
    
    # --------------------------- 4. Resampling ---------------------------
    
    def _fast_resample_and_compute(self, data_structure, current_labels, seed):
        
        # Initialize unique RNG for this worker
        rng = np.random.default_rng(seed)
        
        g0_samples = []
        g1_samples = []
        
        for i in range(self.n_subjects):
            _, cells_list = data_structure[i]
            assigned_group = current_labels[i]
            n_available_cells = len(cells_list)
            
            # -------------- 4.1 Select cells from each subject --------------
       
            if self.replace:
                chosen_cell_indices = rng.choice(n_available_cells, self.n_cells, replace = True)
            else:
                chosen_cell_indices = rng.choice(n_available_cells, self.n_cells, replace = False)
            
            # -------------- 4.2 Sample observations from each selected cell --------------
            
            subject_samples = []
            for c_idx in chosen_cell_indices:
                cell_vals = cells_list[c_idx]
                if self.replace:
                    obs = rng.choice(cell_vals, self.n_obs, replace = True)
                else:
                    obs = rng.choice(cell_vals, self.n_obs, replace = False)
                subject_samples.append(obs)

            subject_samples = np.concatenate(subject_samples)
            
            if assigned_group == 0:
                g0_samples.append(subject_samples)
            else:
                g1_samples.append(subject_samples)
        
        if not g0_samples or not g1_samples:
            return 0
        
        g0_arr = np.concatenate(g0_samples)
        g1_arr = np.concatenate(g1_samples)
        
        # -------------- 4.3 Compute statistic --------------
        
        if self.metric == 'ks':
            return stats.ks_2samp(g0_arr, g1_arr).statistic
        elif self.metric == 'ad':
            with warnings.catch_warnings():
                warnings.simplefilter("ignore")
                return stats.anderson_ksamp([g0_arr, g1_arr]).statistic
            
    # --------------------------- 5. Bootstrap ---------------------------
    
    def run(self, n_resamples = 1000, n_permutations = 1000, n_jobs = -1, random_state = None):
        
        """
        The bootstrap resampling procedure is performed on the observed data to generate a
        distribution of KS or AD statistics and compute its median. Subsequently, a null 
        distribution is obtained by performing the same procedure on data with shuffled 
        subject labels.
        
        The p-value is calculated as the proportion of statistics in the null distribution 
        that are greater than or equal to the median statistic derived from the observed data.
        
        """
        
        if random_state is None:
            master_rng = np.random.default_rng()
        else:
            master_rng = np.random.default_rng(random_state)
            
        original_labels = np.array([item[0] for item in self.optimized_data])
        
        # -------------- 5.1 Observed distribution --------------
        
        print(f"Calculating Observed Distribution ({n_resamples} resamples)...")
        obs_seeds = master_rng.integers(0, 10**9, size = n_resamples)
        
        self.observed_distribution = Parallel(n_jobs = n_jobs)(
            delayed(self._fast_resample_and_compute)(self.optimized_data, original_labels, seed)
            for seed in tqdm(obs_seeds, total = n_resamples, desc = "Observed")
        )
        
        self.observed_distribution = np.array(self.observed_distribution)
        self.observed_stat_median = np.median(self.observed_distribution)
        
        # -------------- 5.2 Null distribution --------------
        
        print(f"Calculating Null Distribution ({n_permutations} permutations)...")
        
        null_seeds = master_rng.integers(0, 10**9, size = n_permutations)
        permuted_label_sets = [master_rng.permutation(original_labels) for _ in range(n_permutations)]
            
        self.null_distribution = Parallel(n_jobs = n_jobs)(
            delayed(self._fast_resample_and_compute)(self.optimized_data, p_labels, seed)
            for p_labels, seed in tqdm(zip(permuted_label_sets, null_seeds), total = n_permutations, desc = "Null")
        )
        
        self.null_distribution = np.array(self.null_distribution)
        
        # -------------- 5.3 P-Value --------------
        
        self.p_value = np.mean(self.null_distribution >= self.observed_stat_median)
        return self.p_value
    
    # --------------------------- 6. Plotting ---------------------------
    
    def plot_results(self, title = None):
        
        if self.p_value is None:
            raise RuntimeError("Run .run() first")
        
        plt.figure(figsize = (10, 5))
        sns.kdeplot(self.null_distribution, fill = True, color = "grey", label = "Null Distribution (Permuted)")
        sns.kdeplot(self.observed_distribution, fill = True, color = "blue", label = "Observed Distribution (Resampled)")
        plt.axvline(self.observed_stat_median, color = 'red', linestyle = '--', label = f'Observed Median: {self.observed_stat_median:.3f}')
        
        metric_name = "KS Statistic" if self.metric == 'ks' else "Anderson-Darling Statistic"
        plt.xlabel(metric_name)
        plt.title(title or f"Nested Permutation Test ({self.metric.upper()})\np-value: {self.p_value:.5f}")
        plt.legend()
        plt.show()
        